[SPARK-45699][BUILD][CORE][SQL][SS][CONNECT][MLLIB][ML][DSTREAM][GRAPHX][K8S][UI] Fixing all compilation warnings related to widening conversions

hannahkamundson opened the pull request:
### What changes were proposed in this pull request?

1. Change the silencing of the widening conversion compilation warnings in the parent `pom.xml` and `SparkBuild` to throw an error
2. All widening conversion compilation warnings were removed. This almost exclusively involved adding `.toDouble` to longs. However, it also involved some `.toFloat` on ints and longs.


### Why are the changes needed?
It allows us to upgrade to Scala 2.13 without adding a bunch of compilation issues. This is removing the following compilation error
```shell
[error] /Users/yangjie01/SourceCode/git/spark-mine-sbt/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1207:60: Widening conversion from Long to Double is deprecated because it loses precision. Write `.toDouble` instead. [quickfixable]
[error] Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=deprecation, site=org.apache.spark.scheduler.TaskSetManager.checkSpeculatableTasks
[error]       foundTasks = checkAndSubmitSpeculatableTasks(timeMs, threshold, customizedThreshold = true)
```


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
No tests were added. 
For every profile (including the base profile), I ran `mvn clean compile test-compile`. I then `grep`ed any lines that had the word `Wide` in it. I determined I was done when no output remained.

Here is a script of what was run:
```shell
mvn clean compile test-compile |& tee output.txt
cat output.txt | grep .*Wide.* |& tee output-widening.txt
mvn clean compile test-compile -Pspark-ganglia-lgpl |& tee output-spark-ganglia-lgpl.txt
cat output-spark-ganglia-lgpl.txt | grep .*Wide.* |& tee output-spark-ganglia-lgpl-widening.txt
mvn clean compile test-compile -Pkinesis-asl |& tee output-kinesis-asl.txt
cat output-kinesis-asl.txt | grep .*Wide.* |& tee output-kinesis-asl-widening.txt
mvn clean compile test-compile -Pdocker-integration-tests |& tee output-docker-integration-tests.txt
cat output-docker-integration-tests.txt | grep .*Wide.* |& tee output-docker-integration-tests-widening.txt
mvn clean compile test-compile -Pyarn \& tee output-yarn.txt
cat output-yarn.txt | grep .*Wide.* |& tee output-yarn-widening.txt
mvn clean compile test-compile -Pkubernetes |& tee output-kubernetes.txt
cat output-kubernetes.txt | grep .*Wide.* |& tee output-kubernetes-widening.txt
mvn clean compile test-compile -Pkubernetes-integration-tests |& tee output-kubernetes-integration-tests.txt
cat output-integration-tests.txt | grep .*Wide.* |& tee output-integration-tests-widening.txt
mvn clean compile test-compile -Phive-thriftserver |& tee output-hive-thriftserver.txt
cat output-thriftserver.txt | grep .*Wide.* |& tee output-thriftserver-widening.txt
mvn clean compile test-compile -Phadoop-cloud |& tee output-hadoop-cloud.txt
cat output-hadoop-cloud.txt | grep .*Wide.* |& tee output-hadoop-cloud-widening.txt
```


### Was this patch authored or co-authored using generative AI tooling?
No

hannahkamundson committed:
[SPARK-45699] Fixing all compilation warnings related to widening conversions

hannahkamundson commented:
@LuciferYang My CICD build is failing because of scalastyle. Could you point me towards a solution to fix this?

hannahkamundson committed:
Fixing scalastyle problem

hannahkamundson committed:
Adding in kubernetes compile warning

LuciferYang commented on core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala:
```diff
@@ -476,7 +476,7 @@ private[spark] object ResourceUtils extends Logging {
         val origTaskAmount = treq.amount
         val taskReqStr = s"${origTaskAmount}/${numParts}"
-        val resourceNumSlots = Math.floor(execAmount * numParts / taskAmount).toInt
+        val resourceNumSlots = Math.floor(execAmount * numParts / taskAmount.toDouble).toInt
```
Before this pr, the result here was a Long value from `execAmount * numParts / taskAmount`. Is there a need to perform `Math.floor` on it? @tgravescs Do you have time to help check this logic? Thanks

tgravescs replied:
it doesn't looks like the floor is needed here. its longs and ints, the other functions called before this take care of doubles.

hannahkamundson replied:
Changed it!

LuciferYang replied:
Thanks @tgravescs

LuciferYang commented on core/src/test/scala/org/apache/spark/benchmark/Benchmark.scala:
```diff
@@ -165,7 +165,8 @@ private[spark] class Benchmark(
     val best = runTimes.min
     val avg = runTimes.sum / runTimes.size
```
how about just make `avg` as a `Double`?

hannahkamundson replied:
I think I'm understanding what you said. Let me know if you meant something different than I posted

LuciferYang replied:
Actually, I hope it's `val avg = runTimes.sum.toDouble / runTimes.size`, then `avg` will be more accurate.

hannahkamundson replied:
I will change it. Is it okay that we are changing the value of average (like you pointed out in the other one)?

hannahkamundson replied:
Done! Let me know if you want me to undo it so the value stays the same

LuciferYang commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala:
```diff
@@ -56,7 +56,7 @@ object EventLogTestHelper {
       desiredSize: Long): Seq[String] = {
     val stringLen = eventStr.getBytes(StandardCharsets.UTF_8).length
-    val repeatCount = Math.floor(desiredSize / stringLen).toInt
+    val repeatCount = Math.floor(desiredSize.toDouble / stringLen).toInt
```
This is the third case related to `Math.floor`. 
`desiredSize / stringLen` is also a long value. Is it necessary to perform `Math.floor` here?

 @HeartSaVioR  Do you have time to help check this logic? Thanks

hannahkamundson replied:
Done!

LuciferYang commented on graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala:
```diff
@@ -321,7 +321,7 @@ class PageRankSuite extends SparkFunSuite with LocalSparkContext {
           0.0
         } else {
-          a * Math.pow(1 - resetProb, vid - source)
+          a * Math.pow(1.0 - resetProb, vid.toDouble - source)
```
nit: `resetProb` is already a double type, it seems unnecessary to explicitly change it to 1.0, right?

hannahkamundson replied:
Done!

LuciferYang commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -150,14 +150,15 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
-      outliers(listWithoutDriver, e => e.totalShuffleWrite) ++
-      outliers(listWithoutDriver, e => e.diskUsed)
+    outliers(listWithoutDriver.filter(_.totalTasks > 0),
+      e => e.totalDuration.toFloat / e.totalTasks) ++
```
Directly change to `e.totalDuration.toFloat` seems to change the result? It might need @dongjoon-hyun  to confirm this change

hannahkamundson replied:
Nice catch! Done.

LuciferYang replied:
hmm... I'm not saying this change is wrong, but it needs confirmation from the developers of the k8s part.

hannahkamundson replied:
Oh ya makes sense! Right now I have it where the value will be the same as it originally was. Let me know if you want me to change that

LuciferYang commented on streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala:
```diff
@@ -67,9 +67,9 @@ private[streaming] object UIUtils {
-    case TimeUnit.NANOSECONDS => milliseconds * 1000 * 1000
-    case TimeUnit.MICROSECONDS => milliseconds * 1000
-    case TimeUnit.MILLISECONDS => milliseconds
+    case TimeUnit.NANOSECONDS => milliseconds * 1000.0 * 1000.0
```
How about unifying it to `milliseconds.toDouble`?

hannahkamundson replied:
Done! Let me know if my change isn't what you meant

LuciferYang reviewed (commented):
https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239

The line above also needs to be deleted in this pr.

Additionally, you can provide an example in the pr description to clarify what the compile warning to be cleaned/fixed looks like.

hannahkamundson commented on common/unsafe/src/test/scala/org/apache/spark/unsafe/types/UTF8StringPropertyCheckSuite.scala:
```diff
@@ -80,7 +80,9 @@ class UTF8StringPropertyCheckSuite extends AnyFunSuite with ScalaCheckDrivenProp
   test("compare") {
     forAll { (s1: String, s2: String) =>
-      assert(Math.signum(toUTF8(s1).compareTo(toUTF8(s2))) === Math.signum(s1.compareTo(s2)))
+      assert(Math.signum {
```
No necessity! It is just what my IDE did when moving it to multiple lines. Is there a style preference in this codebase?

hannahkamundson commented on core/src/main/scala/org/apache/spark/input/FixedLengthBinaryInputFormat.scala:
```diff
@@ -74,7 +74,7 @@ private[spark] class FixedLengthBinaryInputFormat
       recordLength.toLong
     } else {
-      (Math.floor(defaultSize / recordLength) * recordLength).toLong
+      (Math.floor(defaultSize.toDouble / recordLength.toDouble) * recordLength).toLong
```
Done! Removed the `Math.floor(....)`

hannahkamundson committed:
PR feedback: Catching a bug, removing unnecessary Math.floor, and modifying some of the .toDoubles

hannahkamundson commented:
Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)

If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.

hannahkamundson commented:
Wait @LuciferYang I found this [Jira ticket 33805](https://issues.apache.org/jira/browse/SPARK-33805) for the [second line in the pom file](https://github.com/apache/spark/blob/25ee62e19bb00610e79de3284987970d9732195c/pom.xml#L2980). Are you sure I should do that for this Jira ticket? It seems like it may be for another ticket.

(I can help with that ticket too, but it helps me to keep them separate if possible.)

LuciferYang commented:
> Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)
> 
> If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.

We need to clean up the corresponding suppression rule in sbt in this pr to keep the suppression rules for Maven and sbt consistent. It doesn't need to be split into another PR.

LuciferYang commented:
> Wait @LuciferYang I found this [Jira ticket 33805](https://issues.apache.org/jira/browse/SPARK-33805) for the [second line in the pom file](https://github.com/apache/spark/blob/25ee62e19bb00610e79de3284987970d9732195c/pom.xml#L2980). Are you sure I should do that for this Jira ticket? It seems like it may be for another ticket.
> 
> (I can help with that ticket too, but it helps me to keep them separate if possible.)

This part of the work has basically been completed. I will double-check it in the near future(After SPARK-45699 and SPARK-45629 are completed.), and if there are any new related tasks to be done, I will ping you in Jira.

hannahkamundson commented:
> > Thanks for the review @LuciferYang . I fixed the changes and updated the PR. Feel free to give any more feedback :)
> > If it is okay, I will break up the second line [spark/project/SparkBuild.scala](https://github.com/apache/spark/blob/398bff77b1c837aed55f4f10ff1157f7da813570/project/SparkBuild.scala#L239) into another PR because I wasn't planning on that for this one. That'll make it easier for me to keep my changes straight.
> 
> We need to clean up the corresponding suppression rule in sbt in this pr to keep the suppression rules for Maven and sbt consistent. It doesn't need to be split into another PR.

Ohh. In response to my last comments, I was thinking you were saying the line above it in  Maven. I'll change the sbt one!

hannahkamundson committed:
PR feedback: Removing from sbt

hannahkamundson committed:
PR feedback: make avg more accurate

LuciferYang commented on project/SparkBuild.scala:
```diff
@@ -236,7 +236,6 @@ object SparkBuild extends PomBuild {
         // TODO(SPARK-33805): Undo the corresponding deprecated usage suppression rule after
         //  fixed.
         "-Wconf:msg=^(?=.*?method|value|type|object|trait|inheritance)(?=.*?deprecated)(?=.*?since 2.13).+$:s",
-        "-Wconf:msg=^(?=.*?Widening conversion from)(?=.*?is deprecated because it loses precision).+$:s",
```
Sorry, I've reconsidered again. Let's not delete it and change it to 
```
-Wconf:msg=^(?=.*?Widening conversion from)(?=.*?is deprecated because it loses precision).+$:e
```

both sbt and maven

hannahkamundson replied:
Hm, what is the goal of replacing it? My original intent was to get all compilation warnings that say `is deprecated because it loses precision`/`Widening conversion from` gone. It seems like this would be adding a compilation warning for something that isn't there. Thoughts?

hannahkamundson replied:
Deleted my comment as I read it closer! I didn't see the changed `e`. Adding this in!

hannahkamundson committed:
Merge remote-tracking branch 'origin/master' into SPARK-45699

hannahkamundson committed:
PR feedback: THrow an error if any widening conversion compilation warnings are introduced

LuciferYang commented on core/src/main/scala/org/apache/spark/partial/CountEvaluator.scala:
```diff
@@ -57,7 +57,10 @@ private[partial] object CountEvaluator {
+    new BoundedDouble(sum + dist.getNumericalMean,
+      confidence,
+      (sum + low).toDouble,
+      (sum + high).toDouble)
```
nit:
```
new BoundedDouble(
  sum + dist.getNumericalMean, confidence, (sum + low).toDouble, (sum + high).toDouble)
```

hannahkamundson replied:
Done!

LuciferYang commented on core/src/main/scala/org/apache/spark/partial/GroupedCountEvaluator.scala:
```diff
@@ -41,7 +41,10 @@ private[spark] class GroupedCountEvaluator[T : ClassTag](totalOutputs: Int, conf
+      sums.map { case (key, sum) => (
+        key,
+        new BoundedDouble(sum.toDouble, 1.0, sum.toDouble, sum.toDouble)
+      ) }.toMap
```
nit:
```scala
sums.map { case (key, sum) =>
  (key, new BoundedDouble(sum.toDouble, 1.0, sum.toDouble, sum.toDouble))
}.toMap
```

hannahkamundson replied:
Done!

LuciferYang commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1196,15 +1196,17 @@ private[spark] class TaskSetManager(
-      foundTasks = checkAndSubmitSpeculatableTasks(timeMs, threshold, customizedThreshold = true)
+      foundTasks = checkAndSubmitSpeculatableTasks(timeMs,
+        threshold.toDouble,
+        customizedThreshold = true)
```
```scala
foundTasks = checkAndSubmitSpeculatableTasks(
  timeMs, threshold.toDouble, customizedThreshold = true)
```

hannahkamundson replied:
Done!

LuciferYang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -1127,7 +1127,9 @@ case class Range(
           val ndv = math.max(upperIndexPos - lowerIndexPos, 1)
           // Update the lowerIndex and lowerBinValue with upper ones for the next iteration.
-          (upperIndex, upperBinValue, binAr :+ HistogramBin(lowerBinValue, upperBinValue, ndv))
+          (upperIndex,
```
```scala
(upperIndex, upperBinValue,
  binAr :+ HistogramBin(lowerBinValue.toDouble, upperBinValue.toDouble, ndv))
```

hannahkamundson replied:
Done!

LuciferYang commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:
```diff
@@ -74,14 +74,18 @@ class EventTimeWatermarkSuite extends StreamTest with BeforeAndAfter with Matche
+        max = largeValue,
+        min = largeValue,
+        avg = largeValue.toDouble,
+        count = largeValue - 1)
```
```scala
EventTimeStats(
  max = largeValue, min = largeValue, avg = largeValue.toDouble, count = largeValue - 1)
```

hannahkamundson replied:
Done!

LuciferYang reviewed (commented):
The code overall is fine to me, there are some minor formatting issues, I think unnecessary line breaks should be minimized.

hannahkamundson committed:
PR feedback: formatting

hannahkamundson committed:
Merge remote-tracking branch 'origin/master' into SPARK-45699

hannahkamundson commented:
@LuciferYang I updated the formatting! Let me know if you have any other comments before approval :)

LuciferYang reviewed (approved):
+1, LGTM

LuciferYang commented:
![image](https://github.com/apache/spark/assets/1475305/ad03fad1-80e9-45ed-956e-32f3faf31f76)

It seems this is different from the current actual modifications, could you update the PR description?

hannahkamundson commented:
@LuciferYang Updated! How do I merge it in?

LuciferYang commented:
Merged into master for Spark 4.0. Thanks @hannahkamundson and @tgravescs ~
